### ISSUE TITLE
ThisType doesn't enforce type-checking `Object.prototype.call`

### ISSUE THREAD

**Issue by mkermani144** [microsoft/TypeScript#59341]
### 🔎 Search Terms

- `ThisType`
- `ThisType` and `call`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `ThisType`

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ssl=8&ssc=40&pln=1&pc=1#code/MYewdgzgLgBAYiEAuGAVAFgSwqgngBwFMAeAbxgA8UwBXAWwCNCAnGAXwD4YAyGcgM0QAKAJTV6TVmxgBePgCgYMQSFEKlSgPSaYoSCAA2hAHQGQAcyFQsEY7hEBuGNpjEAtG5gtmIZoo3MhFA0zGAw1tjGFA7+bPJx8gggxirGwACGBgZC5LgoAEzsjs467p5gIF7MPsxAA

### 💻 Code

```ts
const Foo: ThisType<{ x: number }> & { foo(): number } = {
  foo() {
    // console.log(this.y); // <-- error
    return this.x;
  }
}

Foo.foo.call({ y: 2 }); // <-- no error
```

### 🙁 Actual behavior

The last line doesn't give an error.

### 🙂 Expected behavior

The last line should give an error.

**Comment by jcalz** [comment]
[`ThisType`](https://www.typescriptlang.org/docs/handbook/utility-types.html#thistypetype) doesn't do what you think it does; it just gives a contextual type to object literals, it doesn't actually constrain it from the outside.  You're probably looking for [a `this` parameter](https://www.typescriptlang.org/docs/handbook/2/classes.html#this-parameters) instead:

```ts
const Foo: { foo(this: { x: number }): number } = {
  foo() {
    console.log(this.y); // <-- error
    return this.x; // okay
  }
}

Foo.foo.call({ y: 2 }); // <-- error
```
[Playground link](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAYiEAuGBvGAzRAKKALASwhXQA8UwBXAWwCMBTAJxgF8BKCmh5lmAXjQAoGJhxshIkaEggANvQB0skAHNchCAoCebANwwA9AZgAeALRmYTRiEbDJjelEqMwMfEQWl9RmCADWAIZa9iyCYYIIIApY0cCBsrLY6FooAEyseobG5pbWtkA)

**Comment by mkermani144** [comment]
Thanks for the quick reply @jcalz.

I know about this parameters, but my point here is that these features may conflict. In addition, when I use `ThisType`, I don't want to duplicate the `this` type for each of the functions in an object.

**Comment by RyanCavanaugh** [comment]
It's not a bug that `ThisType` doesn't do two different things at once

**Comment by mkermani144** [comment]
> It's not a bug that `ThisType` doesn't do two different things at once  

I understand. I'm not necessarily stating it's a bug. Maybe it's more like a new feature. By the way, I think it may benefit developers. Or maybe I'm wrong.
